crates/bcr-ebill-transport/src/handler/bill_invite_handler.rs

```diff
@@ -0,0 +1,343 @@
+                {
+                    error!("Error storing chain events: {e}");
+                }
+            } else {
+                error!("Could not extract chain data for invite event {event:?}");
+            }
+        } else {
+            warn!("Could not decode event to ChainInvite {event:?}");
+        }
+        Ok(())
+    }
+}
+
+impl ServiceTraitBounds for BillInviteEventHandler {}
+
+impl BillInviteEventHandler {
+    pub fn new(
+        transport: Arc<dyn NotificationJsonTransportApi>,
+        processor: Arc<dyn BillChainEventProcessorApi>,
+        chain_event_store: Arc<dyn NostrChainEventStoreApi>,
+    ) -> Self {
+        Self {
+            transport,
+            processor,
+            chain_event_store,
+        }
+    }
+
+    /// Parses chain keys, resolves all Nostr events and builds Nostr chains from it.
+    /// Then decrypts the payloads and parses the block contents. Returns all found chains
+    /// in descending order by chain length.
+    async fn resolve_chain_data(
+        &self,
+        keys: &ChainKeys,
+        chain_id: &str,
+        chain_type: BlockchainType,
+        events: &[nostr_sdk::Event],
+    ) -> Result<Vec<Vec<EventContainer>>> {
+        let keys = BcrKeys::from_private_key(&keys.private_key)?;
+        let mut chains = collect_event_chains(events, chain_id, chain_type, &keys);
+        chains.sort_by_key(|v| Reverse(v.len()));
+        Ok(chains)
+    }
+
+    async fn store_events(
+        &self,
+        chain_id: &str,
+        chain_type: BlockchainType,
+        inserted_chain: Vec<EventContainer>,
+        chains: &[Vec<EventContainer>],
+    ) -> Result<()> {
+        let mut to_insert: HashMap<String, NostrChainEvent> = HashMap::new();
+
+        // first store the inserted valid blocks
+        for (idx, inserted) in inserted_chain.iter().enumerate() {
+            let data = inserted.as_chain_store_event(chain_id, chain_type, idx + 1);
+            to_insert.insert(data.event_id.to_owned(), data);
+        }
+
+        // then store all malicious or invalid blocks with their pretended block height
+        for chain in chains {
+            for (idx, data) in chain.iter().enumerate() {
+                if !to_insert.contains_key(&data.event.id.to_string()) {
+                    let data = data.as_chain_store_event(chain_id, chain_type, idx + 1);
+                    to_insert.insert(data.event_id.to_owned(), data);
+                }
+            }
+        }
+
+        for (_, data) in to_insert {
+            if let Err(e) = self.chain_event_store.add_chain_event(data).await {
+                debug!("Could not store chain event because {e}")
+            }
+        }
+        Ok(())
+    }
+}
+
+#[allow(dead_code)]
+// assumes that events are sorted by timestamp already. Will build up as many chains as needed
+// for the Nostr chain structure. This does not look into the actual blockchain, but will
+// the chain just from Nostr metadata.
+fn collect_event_chains(
```

Comment:
some tests for these operations would probably be good, once things have settled a little :+1:

Reply:
Yes absolutely. I want to add them in the next iteration as I have a refactoring for this in mind concerning the block height.